### PR TITLE
NEXUS-21922 Handle Composite Sites

### PR DESCRIPTION
**Pull request by proteus1121**
Add support composite sites like
https://download.eclipse.org/releases/2019-09/
https://raw.githubusercontent.com/Enide/eclipse-p2-composite-repository/master/

### CODE REVIEW

File: nexus-repository-p2/src/main/java/org/sonatype/nexus/repository/p2/internal/metadata/ArtifactsXmlAbsoluteUrlRemover.java [hunk 1]
```diff
@@ -77,17 +80,18 @@
...
   {
     try {
       Path tempFile = createTempFile("p2-artifacts-" + randomUUID().toString(), "xml");
       // This is required in the case that the input stream is a jar to allow us to extract a single file
       Path artifactsTempFile = createTempFile("jar-entry-" + randomUUID().toString(), "xml");
       try {
-        try (InputStream xmlIn = xmlInputStream(artifact, extension, artifactsTempFile)) {
-          try (OutputStream xmlOut = xmlOutputStream(extension, tempFile)) {
+        try (InputStream xmlIn = xmlInputStream(artifact, ARTIFACTS_XML, extension, artifactsTempFile)) {
+          try (OutputStream xmlOut = xmlOutputStream(ARTIFACTS_XML, extension, tempFile)) {
```
Comment (DreamVan) [inline nexus-repository-p2/src/main/java/org/sonatype/nexus/repository/p2/internal/metadata/ArtifactsXmlAbsoluteUrlRemover.java]: Can be joined to the one 'try' block

File: nexus-repository-p2/src/main/java/org/sonatype/nexus/repository/p2/internal/metadata/ArtifactsXmlAbsoluteUrlRemover.java [hunk 2]
```diff
@@ -122,12 +127,122 @@ public TempBlob removeMirrorUrlFromArtifactsXml(final TempBlob artifact,
...
+      final String file,
+      final String extension)
+  {
+    try {
+      Path tempFile = createTempFile("p2-artifacts-composite-" + randomUUID().toString(), "xml");
+      // This is required in the case that the input stream is a jar to allow us to extract a single file
+      Path artifactsTempFile = createTempFile("jar-entry-" + randomUUID().toString(), "xml");
+      try {
+        try (InputStream xmlIn = xmlInputStream(artifact, file + "." + "xml", extension, artifactsTempFile)) {
+          try (OutputStream xmlOut = xmlOutputStream(file + "." + "xml", extension, tempFile)) {
```
Comment (DreamVan) [inline nexus-repository-p2/src/main/java/org/sonatype/nexus/repository/p2/internal/metadata/ArtifactsXmlAbsoluteUrlRemover.java]: 4 'try' blocks look very odd. Lines 142 and 143 can be combined.

File: nexus-repository-p2/src/main/java/org/sonatype/nexus/repository/p2/internal/proxy/P2ProxyFacetImpl.java [hunk 3]
```diff
@@ -178,6 +179,30 @@ else if (assetKind.equals(AssetKind.ARTIFACT_XML_XZ)) {
         return saveMetadataAsAsset(assetPath, newMetadataContent, payload, assetKind);
       }
     }
+    else if (assetKind.equals(AssetKind.COMPOSITE_ARTIFACTS_JAR)) {
```
Comment (DreamVan) [inline nexus-repository-p2/src/main/java/org/sonatype/nexus/repository/p2/internal/proxy/P2ProxyFacetImpl.java]: Why do not use switch/case?

Comment (proteus1121) [reply nexus-repository-p2/src/main/java/org/sonatype/nexus/repository/p2/internal/proxy/P2ProxyFacetImpl.java]: Make sense, done

File: nexus-repository-p2/src/main/java/org/sonatype/nexus/repository/p2/internal/metadata/ArtifactsXmlAbsoluteUrlRemover.java (hunk 2 above)
Comment (proteus1121) [reply nexus-repository-p2/src/main/java/org/sonatype/nexus/repository/p2/internal/metadata/ArtifactsXmlAbsoluteUrlRemover.java]: Done

File: nexus-repository-p2/src/main/java/org/sonatype/nexus/repository/p2/internal/metadata/ArtifactsXmlAbsoluteUrlRemover.java (hunk 1 above)
Comment (proteus1121) [reply nexus-repository-p2/src/main/java/org/sonatype/nexus/repository/p2/internal/metadata/ArtifactsXmlAbsoluteUrlRemover.java]: Done

File: nexus-repository-p2/src/main/java/org/sonatype/nexus/repository/p2/internal/metadata/ArtifactsXmlAbsoluteUrlRemover.java [hunk 4]
```diff
@@ -77,38 +79,41 @@
...
-                                                  final String extension)
+  private static final String REPOSITORY = "repository";
+
+  public TempBlob removeMirrorUrlFromArtifactsXml(
+      final TempBlob artifact,
+      final Repository repository,
+      final String extension)
   {
     try {
       Path tempFile = createTempFile("p2-artifacts-" + randomUUID().toString(), "xml");
```
Comment (sergeyonofreychuck) [inline nexus-repository-p2/src/main/java/org/sonatype/nexus/repository/p2/internal/metadata/ArtifactsXmlAbsoluteUrlRemover.java]: we can use 
Path tempFile = createTempFile("p2-artifacts", "xml");
the function create a file using the formula "prefix + random_long_no + suffix"

Comment (sergeyonofreychuck) [inline nexus-repository-p2/src/main/java/org/sonatype/nexus/repository/p2/internal/metadata/ArtifactsXmlAbsoluteUrlRemover.java]: Use something like 
  private Path getTemporaryFile(String prefix, String suffix) {
    Path tempFile;
    try {
      tempFile = createTempFile(prefix, suffix);
    }
    catch (IOException e) {
      log.error("Unable to create temporary file");
      throw new RuntimeException(e);
    }
    return tempFile;
  }

**Pull request closed without merging**